crates/ty_python_semantic/src/types/narrow.rs

```diff
@@ -277,48 +276,221 @@ impl ClassInfoConstraintFunction {
     }
 }
 
-type NarrowingConstraints<'db> = FxHashMap<ScopedPlaceId, Type<'db>>;
+/// Represents a TypeGuard-containing disjunct in a Disjunctive Normal Form
+/// (DNF) narrowing constraint.
+///
+/// Such a constraint may optionally have a refinement applied after the type
+/// guard, which is interpreted as being intersected with the type guard.
+///
+/// For example, `(TypeGuardConstraint { typeguard: A, refinement: Some(B) } &
+/// TypeGuardConstraint { typeguard: C, refinement: Some(D) })` evaluates to
+/// `TypeGuardConstraint { typeguard: C, refinement: Some(D) }` because the type guard
+/// in the second conjunct clobbers that in the first.
+#[derive(Hash, PartialEq, Debug, Eq, Clone, salsa::Update, get_size2::GetSize)]
+struct TypeGuardConstraint<'db> {
+    /// If `TypeGuard[T]`, this is `Some(T)`
+    typeguard: Type<'db>,
+    /// If additional constraints are applied _after_ the `TypeGuard`, then they
+    /// go here
+    refinement: Option<Type<'db>>,
+}
+
+impl<'db> TypeGuardConstraint<'db> {
+    /// Evaluate this typeguard constraint to a single type.
+    /// If there's a refinement, it's intersected with the typeguard constraint.
+    fn evaluate_constraint_type(self, db: &'db dyn Db) -> Type<'db> {
+        match self.refinement {
+            Some(refinement) => IntersectionBuilder::new(db)
+                .add_positive(self.typeguard)
+                .add_positive(refinement)
+                .build(),
+            None => self.typeguard,
+        }
+    }
+}
```

Comment:
Did you consider eliminating this struct, making `typeguard_disjuncts` below a `SmallVec` of `Type<'db>'` instead, and eagerly intersecting refinements into the type instead of doing it lazily? Locally that seems to simplify things a bit, and passes all tests.

Reply:
Went ahead and merged with this simplification. If you think it's wrong, let me know (ideally with a code example we now fail) and I'll be happy to make a follow-up PR to restore this.